File: pypy/sweep_sentinel.py
"""Sweep sentinel_sum throughput across a range of p(sentinel).

Rebuilds the 10M-element list at each sentinel fraction and times sentinel_sum,
printing one row per fraction. Run under any interpreter:

    python3 sweep_sentinel.py
    pypy3   sweep_sentinel.py

Pass fractions on the command line to sweep just those (e.g. one per process,
to avoid the JIT compiling the loop once and reusing that trace).
"""

import platform
import random
import sys
import time

from sums import SENTINEL, sentinel_sum

N = 10_000_000
REPEATS = 5
SEED = 1234
DEFAULT_FRACTIONS = [0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.95]
FRACTIONS = [float(a) for a in sys.argv[1:]] or DEFAULT_FRACTIONS


def build_data(n, frac):
    rng = random.Random(SEED)
    data = list(range(n))
    for i in range(n):
        if rng.random() < frac:
            data[i] = SENTINEL
    return data


def best_time(data):
    sentinel_sum(data)  # warm-up for this dataset
    best = float("inf")
    for _ in range(REPEATS):
        start = time.perf_counter()
        sentinel_sum(data)
        best = min(best, time.perf_counter() - start)
    return best


def main():
    print(f"{platform.python_implementation()} {platform.python_version()}  "
          f"(N={N:,}, {REPEATS} reps, best)")
    print(f"  {'p(sent)':>7} {'best (s)':>10} {'M elem/s':>10}")
    for frac in FRACTIONS:
        data = build_data(N, frac)
        best = best_time(data)
        print(f"  {frac:>7.2f} {best:>10.4f} {N / best / 1e6:>10.1f}")


if __name__ == "__main__":
    main()

File: pypy/c/sweep_sentinel.c
/* Sweep sentinel_sum throughput across p(sentinel), mirroring
   sweep_sentinel.{py,js}. Pass fractions as CLI args, else a default set. */
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

#include "bench_util.h"
#include "sums.h"

#define SEED 1234u
#define REPEATS 5
#define N 10000000UL

static double DEFAULT_FRACTIONS[] = {0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
                                     0.6,  0.7, 0.8, 0.9, 0.95};

static int *build_data(size_t n, double frac) {
    rng_t rng;
    rng_seed(&rng, SEED);
    int *data = malloc(n * sizeof *data);
    for (size_t i = 0; i < n; i++) data[i] = (int)i;
    for (size_t i = 0; i < n; i++) {
        if (rng_double(&rng) < frac) data[i] = SENTINEL;
    }
    return data;
}

static double best_time(const int *data, size_t n) {
    double best = 1e30;
    volatile long sink = 0;
    for (int r = 0; r < REPEATS; r++) {
        double s = now_sec();
        sink = sentinel_sum(data, n);
        double e = now_sec() - s;
        if (e < best) best = e;
    }
    (void)sink;
    return best;
}

int main(int argc, char **argv) {
    double *fractions;
    size_t nfrac;
    if (argc > 1) {
        nfrac = (size_t)(argc - 1);
        fractions = malloc(nfrac * sizeof *fractions);
        for (size_t i = 0; i < nfrac; i++) fractions[i] = atof(argv[i + 1]);
    } else {
        fractions = DEFAULT_FRACTIONS;
        nfrac = sizeof DEFAULT_FRACTIONS / sizeof DEFAULT_FRACTIONS[0];
    }

    printf("C (gcc)  N=%lu, %d reps, best\n", N, REPEATS);
    printf("  %7s %10s %10s\n", "p(sent)", "best (s)", "M elem/s");
    for (size_t i = 0; i < nfrac; i++) {
        int *data = build_data(N, fractions[i]);
        double best = best_time(data, N);
        printf("  %7.2f %10.4f %10.1f\n", fractions[i], best, N / best / 1e6);
        free(data);
    }
    return 0;
}
